src/services/ui/src/features/package-actions/shared-components.tsx

```diff
@@ -0,0 +1,262 @@
+        type="reset"
+        disabled={state === "submitting"}
+      >
+        Cancel
+      </Button>
+    </section>
+  );
+};
+
+export const PackageSection = ({ id, type }: { id: string; type: string }) => {
+  return (
+    <section className="flex flex-col my-8 space-y-8">
+      <div>
+        <p>Package ID</p>
+        <p className="text-xl">{id}</p>
+      </div>
+      <div>
+        <p>Type</p>
+        <p className="text-xl">{type}</p>
+      </div>
+    </section>
+  );
+};
+
+export const AdditionalInformation = () => {
+  const form = useFormContext();
+
+  return (
+    <section className="my-4">
+      <h2 className="font-bold text-2xl font-sans mb-2">
+        Additional Information <RequiredIndicator />
+      </h2>
+      <FormField
+        control={form.control}
+        name="additionalInformation"
+        render={({ field }) => (
+          <FormItem>
+            <FormLabel>
+              <p>
+                Add anything else that you would like to share with the State.
+              </p>
+            </FormLabel>
+            <Textarea {...field} className="h-[200px] resize-none" />
+            <FormMessage />
+            <FormDescription>4,000 characters allowed</FormDescription>
+          </FormItem>
+        )}
+      />
+    </section>
+  );
+};
+
+export const ErrorBanner = () => {
+  const form = useFormContext();
+
+  return (
+    <>
+      {Object.keys(form.formState.errors).length !== 0 && (
+        <Alert className="my-6" variant="destructive">
+          Input validation error(s)
+          <ul className="list-disc">
+            {Object.values(form.formState.errors).map(
+              (err, idx) =>
+                err?.message && (
+                  <li className="ml-8 my-2" key={idx}>
+                    {err.message as string}
+                  </li>
+                )
+            )}
+          </ul>
+        </Alert>
+      )}
+    </>
+  );
+};
+
+export const FormLoadingSpinner = () => {
+  const { state } = useNavigation();
+  return state === "submitting" && <LoadingSpinner />;
+};
+
+// Hooks
+
+export const useSubmitForm = () => {
+  const methods = useFormContext();
+  const submit = useSubmit();
+
+  const validSubmission: SubmitHandler<any> = (data, e) => {
+    const formData = new FormData();
+
+    // Append all other data
+    for (const key in data) {
+      if (key !== "attachments") {
+        formData.append(key, data[key]);
+      }
+    }
+    const attachments = data.attachments || {};
+    for (const key in attachments) {
+      attachments[key]?.forEach((file: any, index: number) => {
+        formData.append(`attachments.${key}.${index}`, file as any);
+      });
+    }
+
+    submit(formData, {
+      method: "post",
+      encType: "multipart/form-data",
+    });
+  };
+
+  return {
+    handleSubmit: methods.handleSubmit(validSubmission),
+    formMethods: methods,
+  };
+};
+
+export const useDisplaySubmissionAlert = (header: string, body: string) => {
```

Comment:
I would like to talk more about it, but a good way to think of this hook is like presubmission. It is structuring the data in a way that will adhere to web-standards when it comes to form submissions. Like instead of preventing default behaivors let's use the platform :)

Reply:
So if i'm reading right, this abstracts the effect of showing the success banner? How does this adapt to the need to display a potential error banner while remaining on the form page in the case of a submission failure? @mdial89f brought that case up recently, and we should have a ticket out for it soon so let's make sure this works for that case in the future, too.